lib/palettes_base.hpp

```diff
@@ -0,0 +1,175 @@
+// LICENSE and NOTICE for details. LLNL-CODE-443271.
+//
+// This file is part of the GLVis visualization tool and library. For more
+// information and source code availability see https://glvis.org.
+//
+// GLVis is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef GLVIS_PALETTESBASE_HPP
+#define GLVIS_PALETTESBASE_HPP
+
+#include "gl/types.hpp"
+
+#include <array>
+#include <string>
+#include <iostream>
+#include <fstream>
+#include <memory>
+#include <vector>
+#include <iomanip>
+
+using namespace std;
+
+struct RGBAf
+{
+   float r, g, b, a;
+
+   constexpr RGBAf(float r = 0.0, float g = 0.0, float b = 0.0, float a = 1.0)
+      : r(r), g(g), b(b), a(a) {}
+
+   void Print(ostream& os = cout) const;
+
+   array<float, 4> AsArray() const { return {r, g, b, a}; }
+};
+
+
+class Palette
+{
+public:
+   const string name;
+
+   /// Constructors
+   Palette(const string& name) : name(name) {}
+
+   /// Constructor from vector of RGBAf
+   Palette(const string& name, const vector<RGBAf>& colors)
+      : name(name), colors(colors) {}
+
+   /// Constructor from Nx3 array
+   template <size_t N>
+   Palette(const string& name, const array<array<float,3>,N>& arr);
+
+   /// Constructor from Nx4 array
+   template <size_t N>
+   Palette(const string& name, const array<array<float,4>,N>& arr);
+
+   /// Get size
+   int size() const { return colors.size(); }
+
+   /// Add color to palette
+   void AddColor(float r, float g, float b, float a = 1.0);
+
+   /// Print each color of this palette to a stream
+   void Print(ostream& os = cout) const;
+
+   /// Get color at index i (optionally, use reversed order)
+   RGBAf Color(int i, bool reversed = false) const;
+
+   /// Get all colors as a vector of float arrays
+   vector<array<float,4>> GetData(bool reversed = false) const;
+
+   /// Are any alpha != 1.0?
+   bool IsTranslucent() const;
+
+private:
+   vector<RGBAf> colors;
+};
+
+/// Generates the texture data for a given palette, to be used in OpenGL
+class Texture
+{
+public:
+   /// The palette to create a texture of
+   Palette* const palette;
+   /// Repeat the palette multiple times (negative for reverse); cannot be 0
+   int Nrepeat_;
+   /// Number of colors to discretize with (0 uses the original number of colors)
+   int Ncolors_;
+   /// Is texture smooth or discrete?
+   bool smooth;
+   /// Constructor - generates texture
+   Texture(Palette* palette, int Nrepeat_ = 1, int Ncolors_ = 0,
+           bool smooth = false);
+   /// Texture size
+   int size() const { return texture_data.size(); }
+   /// Get texture data
+   const vector<array<float,4>>& GetData() const { return texture_data; }
+   /// If true, all colors in palette are read in reverse
+   bool IsReversed() const { return Nrepeat_ < 0; }
+   /// Generates the texture data
+   void Generate();
+
+private:
+   int MAX_TEXTURE_SIZE;
```

Comment:
This should be renamed to match the naming convention and also could be static (inited to something invalid and then set by the first instance).

Reply:
I searched the repo and found both camelcase and snakecase for static non-const ints. I set it to snakecase and I also made this static.